### PR TITLE
Fix while-loop graph explosion via WhileLoopPrimitive

### LINKED ISSUE #134: `metal::malloc` resource limit crash on long `lax.scan` / `lax.while_loop`
### Code to reproduce

```python
import jax
import jax.numpy as jnp
from jax import random
def make_lstm_params(key, input_dim, hidden_dim):
    keys = random.split(key, 8)
    # Wi, Wf, Wc, Wo (input projections) + Ui, Uf, Uc, Uo (recurrent projections)
    return tuple(random.normal(k, (input_dim if i < 4 else hidden_dim, hidden_dim)) * 0.01
                 for i, k in enumerate(keys))
def lstm_step(carry, x):
    (h, c), params = carry
    Wi, Wf, Wc, Wo, Ui, Uf, Uc, Uo = params
    i = jax.nn.sigmoid(x @ Wi + h @ Ui)
    f = jax.nn.sigmoid(x @ Wf + h @ Uf)
    g = jnp.tanh(x @ Wc + h @ Uc)
    o = jax.nn.sigmoid(x @ Wo + h @ Uo)
    c_new = f * c + i * g
    h_new = o * jnp.tanh(c_new)
    return ((h_new, c_new), params), h_new
hidden = 64
params = make_lstm_params(random.key(0), 32, hidden)
xs = jnp.ones((10_000, 32))
def loss_fn(params):
    init = ((jnp.zeros(hidden), jnp.zeros(hidden)), params)
    _, ys = jax.lax.scan(lstm_step, init, xs)
    return ys.sum()
jax.grad(loss_fn)(params)  # crashes: metal::malloc resource limit exceeded
```

## Problem
Long-running `lax.scan` / `lax.while_loop` (e.g., LSTM fwd+bwd with ≥10k timesteps) crash with:
```
[MPS ERROR] Exception dispatching stablehlo.dynamic_update_slice: [metal::malloc] Resource limit (499000) exceeded.
[MPS ERROR] stablehlo.while: failed to execute body region
```
## Current state
`HandleWhile` throws `CompileIncompatibleError` when called inside `mx::compile`, forcing the entire outer compilation to fall back to the interpreted MLIR walker. In this path, each loop iteration accumulates MLX graph nodes without evaluation. For long loops, unbounded graph growth exceeds Metal's ~500k allocation limit.

## Approach (I am already preparing a PR)
Replace the `CompileIncompatibleError` throw with a custom `WhileLoopPrimitive` — an `mlx::core::Primitive` subclass that is opaque to `mx::compile` but executes the loop internally:
- loop body wrapped via `mx::compile()` for GPU execution
- The primitive runs on the CPU stream for orchestration only (this does not cause any issues as far as I can tell) — the compiled body dispatches GPU kernels via the default GPU stream. Running on the GPU stream would deadlock (`eval()` inside a GPU eval callback re-enters the GPU scheduler).

### Version and runtime information

```text
jax:     0.9.2
jaxlib:  0.9.2
jax-mps: 0.9.14
Python:  3.13.12
macOS:   26.3.1 (25D2128)
Chip:    Apple M3 Max
Memory:  48 GB
Backend: mps
```

### PR DESCRIPTION
**Pull request by tsumme1**
Fixes #134

## Summary

Fix `metal::malloc` resource limit crash on long `lax.scan` / `lax.while_loop` by introducing a `WhileLoopPrimitive`.

### Problem

The existing `HandleWhile` throws `CompileIncompatibleError` when called inside `mx::compile`, causing the outer compilation to fail and the entire program to fall back to the interpreted MLIR walker. For long-running loops (e.g., LSTM fwd+bwd with ≥10k timesteps), this path accumulates an unbounded MLX computation graph per iteration, eventually exceeding Metal's `malloc` resource limit (499,000 allocations).

### Fix

Replace the `throw` with a custom `WhileLoopPrimitive` (an `mlx::core::Primitive` subclass) that is opaque to `mx::compile` but runs the loop internally with:
- **Compiled body**: the loop body is wrapped via `mx::compile` for efficient GPU execution
- **Per-step `async_eval`**: each iteration is evaluated eagerly, bounding the computation graph
- **Counted-loop detection**: for `lax.scan` patterns (`counter < constant`, body increments by +1), the trip count is extracted at compile time and the condition is never evaluated
- **Dynamic condition support**: general `lax.while_loop` with arbitrary conditions

The primitive runs on the CPU stream for orchestration only — the compiled body dispatches GPU kernels via the default GPU stream. Running on the GPU stream would deadlock (`eval()` inside a GPU eval callback re-enters the GPU scheduler).

### Crash fix (LSTM fwd+bwd, h=64, d=32, t=10000)

| | `main` | This PR |
|---|---|---|
| Result | ❌ `[metal::malloc] Resource limit (499000) exceeded` | ✅ 13.5s |

### GRU training benchmark (fwd + bwd + SGD, T=200, batch=32)

| Config | Unroll | `main` | This PR | Speedup |
|--------|--------|--------|---------|---------|
| h=512 | 1 | 1051 ms | 736 ms | **1.43×** |
| h=512 | 10 | 232 ms | 149 ms | **1.55×** |
| h=512 | 50 | 150 ms | 76 ms | **1.98×** |
| h=512 | 200 (full) | 47 ms | 48 ms | 0.96× |
| h=1024 | 1 | 1654 ms | 1298 ms | **1.27×** |
| h=1024 | 10 | 318 ms | 233 ms | **1.37×** |
| h=1024 | 50 | 180 ms | 110 ms | **1.64×** |
| h=1024 | 200 (full) | 76 ms | 76 ms | 0.99× |
| h=2048 | 1 | 3117 ms | 2558 ms | **1.22×** |
| h=2048 | 10 | 688 ms | 518 ms | **1.33×** |
| h=2048 | 50 | 403 ms | 254 ms | **1.58×** |
| h=2048 | 200 (full) | 184 ms | 176 ms | 1.05× |

**1.2×–2.0× faster** for `unroll < T`. No regression when fully unrolled (where no while-loop is generated).

## How It Works

```
Before (main):
  mx::compile(program_with_while)
    → HandleWhile throws CompileIncompatibleError
    → Entire program falls back to MLIR walker
    → While body graph grows unboundedly per iteration
    → metal::malloc crash at ~500k allocations

After (this PR):
  mx::compile(program_with_while)
    → HandleWhile creates WhileLoopPrimitive (opaque leaf node)
    → Outer compile succeeds
    → On eval, primitive runs loop with compiled body + async_eval per step
    → Bounded memory, no graph explosion
```

### CODE REVIEW

File: tests/configs/control_flow.py [hunk 1]
```diff
@@ -6,6 +6,36 @@
...
+    o = jax.nn.sigmoid(x @ Wo + h @ Uo)
+    c_new = f * c + i * g
+    h_new = o * jnp.tanh(c_new)
+    return ((h_new, c_new), params), h_new
+
+
+def _scan_lstm_loss(params):
+    """Loss function from issue #134 reproducer. Crashes with metal::malloc on main."""
+    hidden = 64
+    xs = jnp.ones((10_000, 32))
```
[the remark this answers is not in the document]
Comment (tsumme1) [reply tests/configs/control_flow.py:33]: That comment was outdated and has been corrected. There probably are ways to get it to be more efficient, but ~10_000 steps is necessary here.

File: src/pjrt_plugin/ops/control_flow.cc [hunk 2]
```diff
@@ -99,7 +295,201 @@ bool HandleWhile(mlir::Operation* op, ValueMap& values, std::vector<mlx::core::a
         return false;
 
     if (ctx.inside_compile) {
-        throw CompileIncompatibleError("stablehlo.while requires eval()");
+        // --- Custom primitive approach ---
+        // Instead of throwing CompileIncompatibleError, we create a
+        // WhileLoopPrimitive that is opaque to mx::compile but runs the
+        // loop with compiled body + per-step eval when eval'd.
+
```
[the remark this answers is not in the document]
Comment (tsumme1) [reply src/pjrt_plugin/ops/control_flow.cc:303]: A test has been added.

File: src/pjrt_plugin/ops/control_flow.cc [hunk 3]
```diff
@@ -4,12 +4,13 @@
 #include <mlx/compile.h>
 #include <mlx/fast.h>
 #include <mlx/linalg.h>
+#include <mlx/primitives.h>
 #include <mlx/random.h>
 #include <mlx/transforms.h>
 
 #include <optional>
 #include <string>
```
[the remark this answers is not in the document]
Comment (tsumme1) [reply src/pjrt_plugin/ops/control_flow.cc:12]: This has been fixed.

File: src/pjrt_plugin/ops/control_flow.cc [hunk 4]
```diff
@@ -91,6 +93,201 @@ bool HandleStablehloReturn(mlir::Operation* op, ValueMap& values,
...
+        }
+    }
+
+    // Constructor for dynamic condition loops (tripCount < 0, with condFn)
+    WhileLoopPrimitive(mlx::core::Stream stream, BodyFn bodyFn, CondFn condFn, size_t nLoopVars,
+                       size_t nExt)
+        : Primitive(stream),
+          compiledBody_(mlx::core::compile(std::move(bodyFn))),
+          compiledCond_(mlx::core::compile(std::move(condFn))),
+          nLoopVars_(nLoopVars),
```
[the remark this answers is not in the document]
Comment (tsumme1) [reply src/pjrt_plugin/ops/control_flow.cc:175]: This is true, but we are essentially just falling back to how the code would currently behave in the same situation. Also, I think this fix for the while loop can be applied to stablehlo.case, which would make all functions compatible with compile, thus eliminating the root problem.

**Comment by tillahoffmann** [comment]
This is great, thank you @tsumme1 !
